core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -418,12 +425,72 @@ public final SegmentsStats segmentsStats() {
                 stats.addTermVectorsMemoryInBytes(guardedRamBytesUsed(segmentReader.getTermVectorsReader()));
                 stats.addNormsMemoryInBytes(guardedRamBytesUsed(segmentReader.getNormsReader()));
                 stats.addDocValuesMemoryInBytes(guardedRamBytesUsed(segmentReader.getDocValuesReader()));
+
+                if (includeSegmentFileSizes) {
+                    // TODO: consider moving this to StoreStats
+                    stats.addFileSizes(getSegmentFileSizes(segmentReader));
+                }
             }
             writerSegmentStats(stats);
             return stats;
         }
     }
 
+    private ImmutableOpenMap<String, Long> getSegmentFileSizes(SegmentReader segmentReader) {
+        Directory directory = null;
+        SegmentCommitInfo segmentCommitInfo = segmentReader.getSegmentInfo();
+        boolean useCompoundFile = segmentCommitInfo.info.getUseCompoundFile();
+        if (useCompoundFile) {
+            try {
+                directory = engineConfig.getCodec().compoundFormat().getCompoundReader(segmentReader.directory(), segmentCommitInfo.info, IOContext.READ);
+            } catch (IOException e) {
+                logger.warn("Error when opening compound reader for Directory [{}] and SegmentCommitInfo [{}]", e,
+                            segmentReader.directory(), segmentCommitInfo);
+
+                return ImmutableOpenMap.of();
+            }
+        } else {
+            directory = segmentReader.directory();
+        }
+
+        assert directory != null;
+
+        String[] files;
+        try {
+            files = directory.listAll();
```

Comment:
> I think instead of directory.listAll(), you could use segmentReader.getSegmentInfo().files()? This should return all file names that this segment uses ...

Hi @mikemccand,
Makes sense. Perhaps we can do:

            if (useCompoundFile) {
                files = directory.listAll();
            } else {
                files = segmentReader.getSegmentInfo().files().toArray(new String[]{});
            }
Otherwise `getSegmentInfo.files()` returns `.cfs`, `.cfe` files and the Compound Directory can't find them when querying `fileLength`